### PR TITLE
Add panic statement support

### LINKED ISSUE #22176: Add panic statement support
Add the `panic` statement support to the new parser.  You can find the syntax in the language specification. 

https://ballerina.io/spec/lang/2020R1/#section_7.24

### PR DESCRIPTION
**Pull request by m36dot**
**Purpose**
Add panic statement support for incremental parser.

**Fixes**: 22176

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaLexer.java [hunk 1]
```diff
@@ -546,6 +546,8 @@ private STToken processIdentifierOrKeyword() {
                 return getSyntaxToken(SyntaxKind.ELSE_KEYWORD);
             case LexerTerminals.WHILE:
                 return getSyntaxToken(SyntaxKind.WHILE_KEYWORD);
+			case LexerTerminals.PANIC:
```
Comment (hasithaa) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaLexer.java]: Not formatted correctly.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 2]
```diff
@@ -158,6 +158,8 @@ public STNode resumeParsing(ParserRuleContext context, STNode... parsedNodes) {
                 return parseElseBody();
             case WHILE_KEYWORD:
                 return parseWhileKeyword();
+			case PANIC_KEYWORD:
```
Comment (hasithaa) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java]: Not formatted correctly.

Comment (hasithaa) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java]: Please check for other files.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaLexer.java (hunk 1 above)
Comment (m36dot) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaLexer.java]: sorry, my editor used tabs,so it looked fine in mine, I ll fix this.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 2 above)
Comment (m36dot) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java]: will do :)

**Comment by SupunS** [comment]
Can you add the recovery logic as well? 

Also we need some test cases.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 3]
```diff
@@ -2833,6 +2837,35 @@ private STNode parseWhileKeyword() {
...
     }
 
+	/**
+     * Parse panic statement.
+     * <code>panic-stmt := panic expression ;</code>
+     * 
+     * @return Panic statement
+     */
+    private STNode parsePanicStatement() {
+        startContext(ParserRuleContext.PANIC_KEYWORD);
```
[the remark this answers is not in the document]
Comment (m36dot) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java]: thanks for pointing this out, I will look in to that.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/syntax/tree/PanicStatement.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    public PanicStatement(STNode node, int position, NonTerminalNode parent) {
+        super(node, position, parent);
+    }
+
+    public Node panicKeyword() {
+        if (panicKeyword != null) {
+            return panicKeyword;
+        }
+
+        panicKeyword = node.childInBucket(0).createFacade(getChildPosition(0), this);
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/syntax/tree/PanicStatement.java]: shouldn't this be `createToken(0);`?

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/statements/PanicStatementTest.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package io.ballerinalang.compiler.parser.test.syntax.statements;
+
+import org.testng.annotations.Test;
+
+/**
+ * Test parsing variable declaration statements.
```
Comment (SupunS) [inline compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/statements/PanicStatementTest.java]: Incorrect doc comment

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 6]
```diff
@@ -2833,6 +2837,35 @@ private STNode parseWhileKeyword() {
...
+        startContext(ParserRuleContext.PANIC_STMT);
+        STNode panicKeyword = parsePanicKeyword();
+        STNode expression = parseExpression();
+        STNode semicolon = parseSemicolon();
+        endContext();
+        return STNodeFactory.createPanicStatement(panicKeyword, expression, semicolon);
+    }
+
+    /**
+     * Parse while-keyword.
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java]: Incorrect doc comment

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/syntax/tree/PanicStatement.java (hunk 4 above)
Comment (m36dot) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/syntax/tree/PanicStatement.java]: yes, I ll fix that

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/statements/PanicStatementTest.java (hunk 5 above)
Comment (m36dot) [reply compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/statements/PanicStatementTest.java]: I ll change that as well

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 6 above)
Comment (m36dot) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java]: I ll change the comments

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/syntax/tree/PanicStatement.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    public PanicStatement(STNode node, int position, NonTerminalNode parent) {
+        super(node, position, parent);
+    }
+
+    public Token panicKeyword() {
+        if (panicKeyword != null) {
+            return panicKeyword;
+        }
+
+        panicKeyword = createToken(0);;
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/syntax/tree/PanicStatement.java]: Additional semicolon

**Pull request merged**